File: node/antchain_node/mempool.py
"""Pending-transaction pool. Admission here is a courtesy pre-check (valid
signature, plausible nonce/balance against the current tip); the real check
happens again at block-validation time, since mempool state can go stale."""
from __future__ import annotations

import threading

from .chain import Blockchain
from .transaction import Transaction


class Mempool:
    def __init__(self):
        self.lock = threading.RLock()
        self._by_hash: dict[str, Transaction] = {}

    def add(self, tx: Transaction, chain: Blockchain) -> bool:
        with self.lock:
            if tx.tx_hash in self._by_hash:
                return False
            if not tx.verify():
                return False
            if chain.balance_of(tx.sender) < tx.amount:
                return False
            self._by_hash[tx.tx_hash] = tx
            return True

    def remove_all(self, txs: list[Transaction]) -> None:
        with self.lock:
            for tx in txs:
                self._by_hash.pop(tx.tx_hash, None)

    def select(self, chain: Blockchain, max_count: int) -> list[Transaction]:
        """Greedily select pending txs whose sender-nonce sequence is
        currently satisfiable, respecting per-sender ordering."""
        with self.lock:
            pending = list(self._by_hash.values())
        by_sender: dict[str, list[Transaction]] = {}
        for tx in pending:
            by_sender.setdefault(tx.sender, []).append(tx)
        for txs in by_sender.values():
            txs.sort(key=lambda t: t.nonce)

        selected: list[Transaction] = []
        next_nonce = {s: chain.next_nonce_for(s) for s in by_sender}
        balances = {s: chain.balance_of(s) for s in by_sender}
        progressed = True
        while progressed and len(selected) < max_count:
            progressed = False
            for sender, txs in by_sender.items():
                for tx in txs:
                    if tx in selected:
                        continue
                    if tx.nonce == next_nonce[sender] and balances[sender] >= tx.amount:
                        selected.append(tx)
                        next_nonce[sender] += 1
                        balances[sender] -= tx.amount
                        progressed = True
                        if len(selected) >= max_count:
                            return selected
                    break  # only the head of each sender's queue can progress
        return selected

    def __len__(self) -> int:
        with self.lock:
            return len(self._by_hash)

File: cpp_node/src/mempool.cpp
#include "mempool.hpp"

#include <algorithm>

namespace antchain_node {

bool Mempool::add(const Transaction& tx, Blockchain& chain) {
    std::string h = tx.tx_hash();
    {
        std::lock_guard<std::mutex> lock(mutex_);
        if (by_hash_.count(h)) return false;
    }
    if (!tx.verify()) return false;
    if (chain.balance_of(tx.sender) < tx.amount) return false;
    std::lock_guard<std::mutex> lock(mutex_);
    by_hash_[h] = tx;
    return true;
}

void Mempool::remove_all(const std::vector<Transaction>& txs) {
    std::lock_guard<std::mutex> lock(mutex_);
    for (const auto& tx : txs) by_hash_.erase(tx.tx_hash());
}

size_t Mempool::size() {
    std::lock_guard<std::mutex> lock(mutex_);
    return by_hash_.size();
}

std::vector<Transaction> Mempool::select(Blockchain& chain, size_t max_count) {
    std::vector<Transaction> pending;
    {
        std::lock_guard<std::mutex> lock(mutex_);
        for (const auto& kv : by_hash_) pending.push_back(kv.second);
    }

    std::map<std::string, std::vector<Transaction>> by_sender;
    for (const auto& tx : pending) by_sender[tx.sender].push_back(tx);
    for (auto& kv : by_sender) {
        std::sort(kv.second.begin(), kv.second.end(),
                  [](const Transaction& a, const Transaction& b) { return a.nonce < b.nonce; });
    }

    std::vector<Transaction> selected;
    std::map<std::string, uint64_t> next_nonce;
    std::map<std::string, double> balances;
    std::map<std::string, size_t> cursor;
    for (const auto& kv : by_sender) {
        next_nonce[kv.first] = chain.next_nonce_for(kv.first);
        balances[kv.first] = chain.balance_of(kv.first);
        cursor[kv.first] = 0;
    }

    bool progressed = true;
    while (progressed && selected.size() < max_count) {
        progressed = false;
        for (auto& kv : by_sender) {
            const std::string& sender = kv.first;
            auto& txs = kv.second;
            size_t& idx = cursor[sender];
            if (idx >= txs.size()) continue;
            const Transaction& tx = txs[idx];
            if (tx.nonce == next_nonce[sender] && balances[sender] >= tx.amount) {
                selected.push_back(tx);
                next_nonce[sender] += 1;
                balances[sender] -= tx.amount;
                idx += 1;
                progressed = true;
                if (selected.size() >= max_count) return selected;
            }
        }
    }
    return selected;
}

} // namespace antchain_node
